flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorState.java

```diff
@@ -81,10 +93,22 @@ public OperatorID getOperatorID() {
         return operatorID;
     }
 
+    public boolean isFullyFinished() {
+        return operatorSubtaskStates.equals(FULLY_FINISHED_STATE);
+    }
+
+    public void markedFullyFinished() {
+        operatorSubtaskStates.clear();
+        FULLY_FINISHED_STATE.forEach(operatorSubtaskStates::put);
+    }
+
     public void putState(int subtaskIndex, OperatorSubtaskState subtaskState) {
         Preconditions.checkNotNull(subtaskState);
+        checkState(!isFullyFinished());
 
-        if (subtaskIndex < 0 || subtaskIndex >= parallelism) {
+        if (subtaskIndex == FULLY_FINISHED_PLACEHOLDER_INDEX) {
+            markedFullyFinished();
```

Comment:
Just to clarify, this branch is for deserialization, right?

Reply:
Yes, this branch is for deserialization. I added comments to this branch

Reply:
Thanks!